reasoner/controller/ConditionController.java

```diff
@@ -22,51 +22,52 @@
+import com.vaticle.typedb.core.reasoner.processor.reactive.PoolingStream;
+import com.vaticle.typedb.core.reasoner.processor.reactive.Reactive.Stream;
 
-import java.util.List;
 import java.util.function.Supplier;
 
-import static com.vaticle.typedb.core.reasoner.processor.reactive.PoolingStream.BufferedFanStream.fanOut;
+import static com.vaticle.typedb.core.reasoner.processor.reactive.PoolingStream.BufferedFanStream.fanInFanOut;
 
-public class ConditionController extends ConjunctionController<
+public class ConditionController extends DisjunctionController<
         Either<ConceptMap, Materialisation>,
-        ConditionController,
-        ConditionController.Processor
+        ConditionController.Processor,
+        ConditionController
         > {
     // Either<> here is just to match the input to ConclusionController, but this class only ever returns ConceptMap
 
     private final Rule.Condition condition;
 
     ConditionController(Driver<ConditionController> driver, Rule.Condition condition, Context context) {
-        super(driver, condition.conjunction(), context);
+        super(driver, condition.disjunction(), context);
         this.condition = condition;
     }
 
     @Override
     protected Processor createProcessorFromDriver(Driver<Processor> processorDriver,
                                                           ConceptMap bounds) {
         return new Processor(
-                processorDriver, driver(), processorContext(), bounds, getPlan(bounds.concepts().keySet()),
+                processorDriver, driver(), processorContext(), disjunction, bounds,
                 () -> Processor.class.getSimpleName() + "(pattern: " + condition.pattern() + ", bounds: " + bounds + ")"
         );
     }
 
     protected static class Processor
-            extends ConjunctionController.Processor<Either<ConceptMap, Materialisation>, Processor> {
+            extends DisjunctionController.Processor<Either<ConceptMap, Materialisation>, Processor> {
 
         private Processor(Driver<Processor> driver, Driver<ConditionController> controller,
-                          Context context, ConceptMap bounds, List<Resolvable<?>> plan,
+                          Context context, ResolvableDisjunction disjunction, ConceptMap bounds,
                           Supplier<String> debugName) {
-            super(driver, controller, context, bounds, plan, debugName);
+            super(driver, controller, context, disjunction, bounds, debugName);
         }
 
         @Override
-        public void setUp() {
-            setHubReactive(fanOut(this));
-            new CompoundStream(this, plan, bounds)
-                    .map(Either::<ConceptMap, Materialisation>first)
-                    .buffer().registerSubscriber(hubReactive());
+        Stream<Either<ConceptMap, Materialisation>, Either<ConceptMap, Materialisation>> getOrCreateHubReactive(Stream<ConceptMap,ConceptMap> fanIn) {
```

Comment:
ill take this is right, too much context to learn in one quick review :)

Reply:
Sure, but the context is:
`fanIn` is the first stream which connects to (all the) upstream controllers - connected to the 'input ports'.
This method fills in the black-box with the pipeline of streams all the way till the 'hubReactive', which is what will get plugged into the output port.